receiver/elasticapmreceiver/internal/sharedcomponent/sharedcomponent.go

```diff
@@ -0,0 +1,168 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package sharedcomponent exposes functionality for components
+// to register against a shared key, such as a configuration object, in order to be reused across signal types.
+// This is particularly useful when the component relies on a shared resource such as os.File or http.Server.
+package sharedcomponent // import "github.com/elastic/opentelemetry-collector-components/internal/sharedcomponent"
+
+import (
+	"context"
+	"slices"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componentstatus"
+)
+
+func NewMap[K comparable, V component.Component]() *Map[K, V] {
+	return &Map[K, V]{
+		components: map[K]*Component[V]{},
+	}
+}
+
+// Map keeps reference of all created instances for a given shared key such as a component configuration.
+type Map[K comparable, V component.Component] struct {
+	lock       sync.Mutex
```

Comment:
Hmm, I am not sure if we require a mutex here. Based on my understanding, the component initialization is done sequentially from the same goroutine. Do you have a reason for adding this?

The [upstream shared component](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/sharedcomponent/sharedcomponent.go) also doesn't do this. Also, should we just copy that internal package?

Reply:
There's actually two implementations, I copied the one from the core collector repo: https://github.com/open-telemetry/opentelemetry-collector/tree/main/internal/sharedcomponent

Reply:
Ah, TIL. Interestingly, the mutex was added as part of the effort to expose the sharedcomponent API: https://github.com/open-telemetry/opentelemetry-collector/pull/9157